external-crates/move/move-compiler/src/editions/mod.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module controls feature gating and breaking changes in new editions of the source language
+
+use std::{fmt::Display, str::FromStr};
+
+use crate::{diag, shared::CompilationEnv};
+use move_ir_types::location::*;
+use serde::{Deserialize, Serialize};
+
+//**************************************************************************************************
+// types
+//**************************************************************************************************
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum Edition {
+    Legacy,
+    E2024(EditionRelease),
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug, PartialOrd, Ord)]
+pub enum EditionRelease {
+    Final,
+    Beta,
+    Alpha,
+}
+
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum FeatureGate {
+    LetMut,
+    ReceiverSyntax,
+    MacroFun,
+    UnderscoreType,
```

Comment:
nit: Do we need to add these up-front, or can we add them as we add the features?

Reply:
We can add them, I only put them there as a sort of example/placeholder. I will remove if that is easier